src/vm/methodtable.cpp

```diff
@@ -1659,73 +1657,101 @@ BOOL MethodTable::CanCastToNonVariantInterface(MethodTable *pTargetMT)
 }
 
 //==========================================================================================
-TypeHandle::CastResult MethodTable::CanCastToInterfaceNoGC(MethodTable *pTargetMT)
+BOOL MethodTable::CanCastToClassOrInterface(MethodTable* pTargetMT, TypeHandlePairList* pVisited)
 {
     CONTRACTL
     {
-        NOTHROW;
-        GC_NOTRIGGER;
-        MODE_ANY;
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
         INSTANCE_CHECK;
         PRECONDITION(CheckPointer(pTargetMT));
-        PRECONDITION(pTargetMT->IsInterface());
+        PRECONDITION(!pTargetMT->IsArray());
         PRECONDITION(IsRestored_NoLogging());
     }
-    CONTRACTL_END
+        CONTRACTL_END
 
-    if (!pTargetMT->HasVariance() && !IsArray() && !HasTypeEquivalence() && !pTargetMT->HasTypeEquivalence())
-    {
-        return CanCastToNonVariantInterface(pTargetMT) ? TypeHandle::CanCast : TypeHandle::CannotCast;
-    }
-    else
+        BOOL result = pTargetMT->IsInterface() ?
+                                    CanCastToInterface(pTargetMT, pVisited) :
+                                    CanCastToClass(pTargetMT, pVisited);
+
+    // We only consider type-based conversion rules here.
+    // Therefore a negative result cannot rule out convertibility for ICastable and COM objects
+    if (result || !(pTargetMT->IsInterface() && (this->IsComObjectType() || this->IsICastable())))
     {
-        // We're conservative on variant interfaces and types with equivalence
-        return TypeHandle::MaybeCast;
+        CastCache::TryAddToCache(this, pTargetMT, (BOOL)result);
     }
+
+    return result;
 }
 
 //==========================================================================================
-TypeHandle::CastResult MethodTable::CanCastToClassNoGC(MethodTable *pTargetMT)
+BOOL MethodTable::ArraySupportsBizarreInterface(MethodTable * pInterfaceMT, TypeHandlePairList* pVisited)
```

Comment:
There are two `ArraySupportsBizarreInterface` methods now. The other one does not have `TypeHandlePairList* pVisited` argument. Does this one need `TypeHandlePairList* pVisited` argument?